libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1819,18 +1821,54 @@ string YulUtilFunctions::copyValueArrayStorageToStorageFunction(ArrayType const&
 				if gt(length, 0xffffffffffffffff) { <panic>() }
 				<resizeArray>(dst, length)
 
-				let srcPtr := <srcDataLocation>(src)
-
-				let dstPtr := <dstDataLocation>(dst)
+				let srcSlot := <srcDataLocation>(src)
+				let dstSlot := <dstDataLocation>(dst)
 
 				let fullSlots := div(length, <itemsPerSlot>)
-				let i := 0
-				for { } lt(i, fullSlots) { i := add(i, 1) } {
-					sstore(add(dstPtr, i), <maskFull>(sload(add(srcPtr, i))))
-				}
-				let spill := sub(length, mul(i, <itemsPerSlot>))
+				let spill := sub(length, mul(fullSlots, <itemsPerSlot>))
+				let totalSlots := fullSlots
 				if gt(spill, 0) {
-					sstore(add(dstPtr, i), <maskBytes>(sload(add(srcPtr, i)), mul(spill, <bytesPerItem>)))
+					totalSlots := add(totalSlots, 1)
+				}
+
+				let srcSlotValue := sload(srcSlot)
+				<?srcReadMultiPerSlot>let srcItemIndexInSlot := 0</srcReadMultiPerSlot>
+
+				for { let i := 0 } lt(i, totalSlots) { i := add(i, 1) } {
+					let dstSlotValue := 0
+					<?sameType>
+						switch lt(i, fullSlots)
```

Comment:
Wouldn't it be better to move the last (i.e. the non-full) slot to after the loop?